src/coreComponents/physicsSolvers/multiphysics/CoupledSolver.hpp

```diff
@@ -0,0 +1,304 @@
+  {
+    forEachArgInTuple( m_solvers, [&]( auto & solver, auto )
+    {
+      solver->assembleSystem( time_n, dt, domain, dofManager, localMatrix, localRhs );
+    } );
+
+    assembleCouplingTerms( time_n, dt, domain, dofManager, localMatrix, localRhs );
+  }
+
+  virtual void
+  applySystemSolution( DofManager const & dofManager,
+                       arrayView1d< real64 const > const & localSolution,
+                       real64 const scalingFactor,
+                       DomainPartition & domain ) override
+  {
+    forEachArgInTuple( m_solvers, [&]( auto & solver, auto )
+    {
+      solver->applySystemSolution( dofManager, localSolution, scalingFactor, domain );
+    } );
+  }
+
+  virtual void
+  updateState( DomainPartition & domain ) override
+  {
+    forEachArgInTuple( m_solvers, [&]( auto & solver, auto )
+    {
+      solver->updateState( domain );
+    } );
+  }
+
+  virtual void
+  resetStateToBeginningOfStep( DomainPartition & domain ) override
+  {
+    forEachArgInTuple( m_solvers, [&]( auto & solver, auto )
+    {
+      solver->resetStateToBeginningOfStep( domain );
+    } );
+  }
+
+
+  virtual real64
+  solverStep( real64 const & time_n,
+              real64 const & dt,
+              int const cycleNumber,
+              DomainPartition & domain ) override
+  {
+    GEOSX_MARK_FUNCTION;
+
+    real64 dt_return = dt;
+
+    // setup the coupled linear system
+    setupSystem( domain, m_dofManager, m_localMatrix, m_rhs, m_solution );
+
+    // setup reservoir and well systems
+    implicitStepSetup( time_n, dt, domain );
+
+    // currently the only method is implicit time integration
+    dt_return = nonlinearImplicitStep( time_n, dt, cycleNumber, domain );
+
+    // complete time step in reservoir and well systems
+    implicitStepComplete( time_n, dt_return, domain );
+
+    return dt_return;
+  }
+
+
+  virtual real64
+  calculateResidualNorm( DomainPartition const & domain,
+                         DofManager const & dofManager,
+                         arrayView1d< real64 const > const & localRhs ) override
+  {
+    real64 norm = 0.0;
+    forEachArgInTuple( m_solvers, [&]( auto & solver, auto )
+    {
+      real64 const singlePhysicsNorm = solver->calculateResidualNorm( domain, dofManager, localRhs );
+      norm += singlePhysicsNorm * singlePhysicsNorm;
+    } );
+    return sqrt( norm );
+  }
+
+  virtual void
+  applyBoundaryConditions( real64 const time_n,
+                           real64 const dt,
+                           DomainPartition & domain,
+                           DofManager const & dofManager,
+                           CRSMatrixView< real64, globalIndex const > const & localMatrix,
+                           arrayView1d< real64 > const & localRhs ) override
+  {
+    forEachArgInTuple( m_solvers, [&]( auto & solver, auto )
+    {
+      solver->applyBoundaryConditions( time_n, dt, domain, dofManager, localMatrix, localRhs );
+    } );
```

Comment:
I think the order doesn't matter....is there a scenario where it would?

Reply:
Maybe in corner cases (say, if a node is shared by boundary faces where we apply different boundary conditions), the order of application of the boundary conditions could accidentally matter (before and after the PR). To fix that, we could implement an xml-decided order for the application of the boundary conditions.